### PR TITLE
Add plugin version

### LINKED ISSUE #6823: Log plugin version
Since we've broken lockstep between Certbot's different components, your Certbot version and the version of a separately packaged plugin might not match. This is fine, but I think it's valuable to be able to easily see what version of a plugin is being used for the purposes of debugging.

One instance where this would have been valuable is users who upgraded their version of Certbot to workaround the TLS-SNI-01 removal, but hadn't upgraded the Apache/Nginx plugin. In this case, they were reporting that they had a recent version of Certbot but it continued to use TLS-SNI-01. Being able to quickly tell the plugins are out-of-date would have been useful.

Currently, I don't think it's possible to get this information out of Certbot and instead you have to ask whatever package manager you used to install the client. I personally think we should add the version string to Certbot's logs. I'd also be interested in seeing it added to the user agent string so we could easily see things like how many people are using a problematic version of a plugin. A third place we could add it is `certbot plugins` output.

I'm not immediately sure the best way to go about this with our current plugin API but I'm happy to brainstorm with someone if they're interested in working on this.

@maximillianh, I think fixing this shouldn't be too hard and is valuable so I'm adding it to our next milestone but I'm happy to revisit that choice.

### PR DESCRIPTION
**Pull request by osirisinferi**
Fixes #6823 

This is a second version of my attempt to add plugin versions to the certbot log. My [first version](https://github.com/certbot/certbot/commit/9661d575c05ef8566571c951383612079c08a19a) was "clean" in terms of number of added lines (few), but by doing so the function `find_init()` was ran twice. Not necessary IMO, so I wrote this second version. A few more lines and if/else et cetera, but better I think.

It seems the class `pkg_resources.EggInfoDistribution` (the type of `entry_point.dist`) contains a version number after all.

@bmw How would you envision the certbot UA string? Currently it's:

```
ua = ("CertbotACMEClient/{0} ({1}; {2}{8}) Authenticator/{3} Installer/{4} ({5}; flags: {6}) Py/{7}")
```

Where to put the version number? Between parenthesis behind the plugin name with a space in between? Or without parenthesis and separated by a hyphen? Examples respectively:

```
ua = ("CertbotACMEClient/{0} ({1}; {2}{8}) Authenticator/{3} ({9}) Installer/{4} ({10}) ({5}; flags: {6}) Py/{7}")

ua = ("CertbotACMEClient/{0} ({1}; {2}{8}) Authenticator/{3}-{9} Installer/{4}-{10} ({5}; flags: {6}) Py/{7}")
```

Or perhaps something else altogether. I have no clue how your log parser is operating.

### CODE REVIEW

File: certbot/certbot/_internal/plugins/disco.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ class PluginEntryPoint(object):
 
     def __init__(self, entry_point, with_prefix=False):
         self.name = self.entry_point_to_plugin_name(entry_point, with_prefix)
+        self.version = entry_point.dist.version
```
Comment (osirisinferi) [inline certbot/certbot/_internal/plugins/disco.py:56]: Is this line OK here? Or should it be after `entry_point.load()`?

**Comment by bmw** [comment]
> @bmw How would you envision the certbot UA string?

I don't have that strong of an opinion, but I think I personally prefer appending it to the end of the plugin name after an underscore like:
```
ua = ("CertbotACMEClient/{0} ({1}; {2}{8}) Authenticator/{3}_{9} Installer/{4}_{10} ({5}; flags: {6}) Py/{7}")
```
My reasons are I think appending it to the plugin name will affect existing queries on logs less as they're essentially regexes that often split on whitespace between fields and I prefer an underscore over a hyphen since our DNS plugins contain a hyphen in their name, but none of them contain an underscore.

I'll also flag that while this is almost certainly fine, we need to get a lawyer at EFF to sign off on us changing the information we collect in our UA string. I've asked @maximillianh to help with that.

**Comment by osirisinferi** [comment]
Interesting, the `test linux-integration` check wasn't failing when I ran the full local test suite through `tox --skip-missing-interpreters`. But seems legit, checking it out..

**Comment by osirisinferi** [comment]
I'm not very pleased by that last commit.. With functions as `unregister` and such there are no plugins selected, so there also isn't a version present. As `config.auth_version` and `config.inst_version` are initialized only through `record_chosen_plugins` in `selection.py`, certbot fails at `unregister`..

I'd rather see the variables `config.auth_version` and `config.inst_version` initialized somewhere to `None`, but I couldn't figure out what a good place was to do that. So I made a nasty work-around in ec94d65 :stuck_out_tongue: Thoughts please!

**Comment by osirisinferi** [comment]
Finally fixed the integration test..

**Comment by adferrand** [comment]
@bmw to you want to continue on that PR since you started to comment, or do you want me to take over?

**Comment by bmw** [comment]
You taking over would be great.

Keep in mind though that as I flagged at https://github.com/certbot/certbot/pull/8334#issuecomment-701527368, we need to get legal review on changes to our user-agent string and I never heard back from Max about this.

I just reached out to a lawyer directly myself and I'll update this issue if I hear anything. With this in mind, you could start reviewing this PR if you wanted, however, it should not be merged without hearing back from a lawyer at EFF.

**Comment by bmw** [comment]
Lawyers say this is fine so by all means take over if you're interested in reviewing this Adrien.

**Comment by osirisinferi** [comment]
Any idea when this could be reviewed?

I almost tried to add a version number to `certbot plugins` **AGAIN** until I remembered this PR and noticed it wasn't merged yet. I assumed it was already merged and I thought I didn't add it in `certbot plugins` (but only to the user agent), but I apparently actually *did* add it here. So no need to do the work twice.

**Pull request closed without merging**